middleware/markdown/markdown.go

```diff
@@ -103,53 +108,45 @@ func (md Markdown) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 		fpath = idx
 	}
 
-	// If supported extension, process it
-	if _, ok := cfg.Extensions[path.Ext(fpath)]; ok {
-		f, err := md.FileSys.Open(fpath)
-		if err != nil {
-			if os.IsPermission(err) {
-				return http.StatusForbidden, err
-			}
-			return http.StatusNotFound, nil
-		}
-
-		fs, err := f.Stat()
-		if err != nil {
-			return http.StatusNotFound, nil
-		}
-		lastModTime = latest(lastModTime, fs.ModTime())
-
-		body, err := ioutil.ReadAll(f)
-		if err != nil {
-			return http.StatusInternalServerError, err
-		}
+	// If not supported extension, pass on it
+	if _, ok := cfg.Extensions[path.Ext(fpath)]; !ok {
+		return md.Next.ServeHTTP(w, r)
+	}
 
-		ctx := middleware.Context{
-			Root: md.FileSys,
-			Req:  r,
-			URL:  r.URL,
-		}
-		html, err := cfg.Markdown(fpath, body, dirents, ctx)
-		if err != nil {
-			return http.StatusInternalServerError, err
-		}
+	// At this point we have a supported extension/markdown
+	f, err := md.FileSys.Open(fpath)
+	switch {
+	case err == nil: // nop
+	case os.IsPermission(err):
+		return http.StatusForbidden, err
+	case os.IsExist(err):
+		return http.StatusNotFound, nil
+	default: // did we run out of FD?
+		return http.StatusInternalServerError, err
+	}
+	defer f.Close()
 
-		// TODO(weingart): move template execution here, something like:
-		//
-		// html, err = md.execTemplate(cfg, html, ctx)
-		// if err != nil {
-		// 	return http.StatusInternalServerError, err
-		// }
+	if fs, err := f.Stat(); err != nil {
+		return http.StatusGone, nil
+	} else {
+		lastModTime = latest(lastModTime, fs.ModTime())
+	}
 
-		middleware.SetLastModifiedHeader(w, lastModTime)
-		if r.Method == http.MethodGet {
-			w.Write(html)
-		}
-		return http.StatusOK, nil
+	ctx := middleware.Context{
+		Root: md.FileSys,
+		Req:  r,
+		URL:  r.URL,
+	}
+	html, err := cfg.Markdown(title(fpath), f, dirents, ctx)
+	if err != nil {
+		return http.StatusInternalServerError, err
 	}
 
-	// Didn't qualify to serve as markdown; pass-thru
-	return md.Next.ServeHTTP(w, r)
+	middleware.SetLastModifiedHeader(w, lastModTime)
+	if r.Method == http.MethodGet {
+		w.Write(html)
+	}
```

Comment:
I think I like:
```go
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.Header().Set("Content-Length", strconv.FormatInt(len(html), 10))
	middleware.SetLastModifiedHeader(w, lastModTime)
	if r.Method == http.MethodGet {
		w.Write(html)
	}
	return http.StatusOK, nil
```
better.  Always set content length and type, only write contents on get.